### PR TITLE
[hail] Introduce proper ptype management to aggregators

### PR DESCRIPTION
**Pull request by tpoterba**
Stacked on #7698

This is the last stacked PR I'm assigning to you this week, I promise!

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala [hunk 1]
```diff
@@ -330,6 +330,13 @@ object IRParser {
     struct_field(type_expr(env))(it)
   }
 
+  def ptype_exprs(env: TypeParserEnvironment)(it: TokenIterator): Array[PType] = {
+    punctuation(it, "(")
+    val types = repUntil(it, type_expr(env), PunctuationToken(")")).map(_.physicalType)
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/Parser.scala]: I'm a little bit hung up on this line, as I'm not sure quite where this is going---is the intention to expose PTypes in python?

If not, how are we going to handle the IR serialization between python and Scala? Always convert to a default PType and then change if necessary?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: no ptype serialization between scala and python -- but I guess we've been trying to preserve the ability to do round trips with Pretty, right?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: I'll write a ptype serializer/deserializer separately, I think.

File: hail/src/main/scala/is/hail/expr/ir/AggOp.scala [hunk 2]
```diff
@@ -1,16 +1,36 @@
...
   op: AggOp,
   initOpArgs: Seq[Type],
   seqOpArgs: Seq[Type],
-  nested: Option[Seq[AggSignature]]) {
-  lazy val returnType: Type = agg.Extract.getType(this)
+  nested: Option[Seq[AggSignature]] = None) {
+  lazy val returnType: Type = agg.Extract.getResultType(this)
+  def toPhysical(initOpTypes: Seq[PType], seqOpTypes: Seq[PType]): PhysicalAggSignature = {
+    assert(nested.isEmpty)
+    (initOpTypes, initOpArgs).zipped.foreach { case (pt, t) => pt.virtualType == t }
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/AggOp.scala]: yep, good catch

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala (hunk 1 above)
Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: So if I understand this correctly---the Python aggregations are going to create AggSignatures that use virtual types, and then Scala will convert them to ptypes and change as necessary?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: this is roughly right, but scala doesn't convert them to ptypes when it parses -- it converts them to ptypes in Extract (which should be a lowering pass soon), when we know that InferPtypes has already been run. 

The SeqOp node, for instance is not in Python, only scala. The physical signatures are only on the lower-level agg nodes.

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: ah, that makes more sense. Thanks!

**Pull request merged**